pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -108,53 +108,31 @@ func (r *reconciler) add(rule *CompletedRule) error {
 	}()
 	// TODO: Update types.PolicyRule to use Antrea Direction type.
 	var direction networkingv1.PolicyType
-	// TODO: Differentiate rule that match everything and rule that match nothing.
-	// nil slice should match everything and empty slice should match nothing.
-	var from, to []types.Address
+	var from, exceptFrom, to, exceptTo []types.Address
 	if rule.Direction == v1beta1.DirectionIn {
 		direction = networkingv1.PolicyTypeIngress
-
-		if rule.FromAddresses != nil || rule.From.IPBlocks != nil {
-			from = make([]types.Address, 0, len(rule.FromAddresses)+len(rule.From.IPBlocks))
-		}
-		for a := range rule.FromAddresses {
-			from = append(from, openflow.NewIPAddress(net.ParseIP(a)))
-		}
-		// TODO: handle except field.
-		for _, b := range rule.From.IPBlocks {
-			from = append(from, openflow.NewIPNetAddress(antreaIPNetToIPNet(b.CIDR)))
-		}
-
+		from, exceptFrom = ipsToOFAddresses(rule.FromAddresses, rule.From.IPBlocks)
 		to = r.podsToOFPortAddresses(rule.Pods)
 	} else {
 		direction = networkingv1.PolicyTypeEgress
-
 		from = r.podsToIPAddresses(rule.Pods)
-
-		if rule.ToAddresses != nil || rule.To.IPBlocks != nil {
-			to = make([]types.Address, 0, len(rule.ToAddresses)+len(rule.To.IPBlocks))
-		}
-		for a := range rule.ToAddresses {
-			to = append(to, openflow.NewIPAddress(net.ParseIP(a)))
-		}
-		// TODO: handle except field.
-		for _, b := range rule.To.IPBlocks {
-			to = append(to, openflow.NewIPNetAddress(antreaIPNetToIPNet(b.CIDR)))
-		}
+		to, exceptTo = ipsToOFAddresses(rule.ToAddresses, rule.To.IPBlocks)
 	}
 
 	// TODO: Update types.PolicyRule to use Antrea Service type.
 	services := servicesToNetworkPolicyPort(rule.Services)
 
 	ofRule := &types.PolicyRule{
-		ID:        ofID,
-		Direction: direction,
-		From:      from,
-		To:        to,
-		Service:   services,
+		ID:         ofID,
+		Direction:  direction,
+		From:       from,
+		ExceptFrom: exceptFrom,
+		To:         to,
+		ExceptTo:   exceptTo,
+		Service:    services,
 	}
 
-	klog.V(2).Infof("Installing ofRule %d (%v, %d From, %d To, %d Service)",
+	klog.V(2).Infof("Installing ofRule %d (Direction: %v, From: %d, To: %d, Service: %d)",
```

Comment:
do you also want to log the Except fields here?

Reply:
Good catch, I forgot this patch is adding this support ;)

Reply:
Done